### PR TITLE
Arbiter for internal mode

### PR DESCRIPTION
**Pull request by cloudbehl**
Screenshot 2020-12-14 at 5 05 19 PM
Signed-off-by: Ankush Behl <cloudbehl@gmail.com>

### CODE REVIEW

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/node-list.tsx [hunk 1]
```diff
@@ -36,10 +36,16 @@ const getRows: GetRows = (
...
+  selectedArbiterZone,
 ) => {
   const { data } = componentProps;
 
-  const filteredData = nodesWithoutTaints(data);
+  const filteredData = nodesWithoutTaints(data).filter((filteredNode) => {
+    return enableStretchCluster && selectedArbiterZone
+      ? getZone(filteredNode) !== selectedArbiterZone
+      : true;
+  });
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/node-list.tsx:48]:
```suggestion
  const filteredData = nodesWithoutTaints(data).filter((filteredNode) =>  enableStretchCluster && selectedArbiterZone
      ? getZone(filteredNode) !== selectedArbiterZone
      : true);
  
```

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard.tsx [hunk 2]
```diff
@@ -92,7 +92,7 @@ export const CreateInternalCluster: React.FC<CreateInternalClusterProps> = ({ ma
       component: (
         <ReviewAndCreate state={state} errorMessage={errorMessage} inProgress={inProgress} />
       ),
-      enableNext: hasEnabledCreateStep,
+      enableNext: hasEnabledCreateStep && !state.isArbiterValid,
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard.tsx:95]:
```suggestion
      enableNext: hasEnabledCreateStep
```
Why not keep it as a part of `hasEnabledCreateStep`?

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx [hunk 3]
```diff
@@ -42,6 +59,110 @@ const validate = (scName, enableMinimal, enableFlexibleScaling): ValidationType[
   return validations;
 };
 
+const isArbiterDisabled = (nodesData: NodeKind[], validateSelectedNodes?: boolean) => {
+  const validation = {
+    nodesLen: validateSelectedNodes ? 4 : 5,
+    zoneLen: validateSelectedNodes ? 2 : 3,
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx:65]:
```suggestion
    nodesLength: validateSelectedNodes ? 4 : 5,
    zoneLength: validateSelectedNodes ? 2 : 3,
```
Let's just go with the full names.

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx [hunk 4]
```diff
@@ -42,6 +59,110 @@ const validate = (scName, enableMinimal, enableFlexibleScaling): ValidationType[
...
+  };
+  const uniqZones: Set<string> = new Set(nodesData.map((node) => getZone(node)));
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx:67]:
```suggestion
  const uniqZones: Set<string> = new Set(nodesData.map(getZone));
```

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx [hunk 5]
```diff
@@ -42,6 +59,110 @@ const validate = (scName, enableMinimal, enableFlexibleScaling): ValidationType[
...
+  if (nodesData.length < validation.nodesLen) return true;
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx:68]:
```suggestion
  if (nodesData.length < validation.nodesLen) return true;
  const uniqZones: Set<string> = new Set(nodesData.map((node) => getZone(node)));
```
We can return early.

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx [hunk 6]
```diff
@@ -42,6 +59,110 @@ const validate = (scName, enableMinimal, enableFlexibleScaling): ValidationType[
...
+  if (uniqZones.size !== validation.zoneLen) return true;
+  const zonePerNode = countNodesPerZone(nodesData);
+  let count = 0;
+  Object.keys(zonePerNode).forEach((zone) => {
+    if (zonePerNode[zone] >= 2) count += 1;
+  });
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx:74]:
```suggestion
const count = Object.keys(zonePerNode).reduce((acc, curr) => {
    if(zonesPerNode[zone] > =2) acc += 1;
    return acc;
    }, 0);
```
Let's use another name for `count`. Something that would give more insight on actually what we are counting.

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx [hunk 7]
```diff
@@ -42,6 +59,110 @@ const validate = (scName, enableMinimal, enableFlexibleScaling): ValidationType[
...
+  dispatch,
+}) => {
+  const { t } = useTranslation();
+  const { enableStretchCluster, selectedArbiterZone, nodes } = state;
+  const [zonesOption, setZonesOptions] = React.useState({});
+  const [showInfoAlert, setShowInfoAlert] = React.useState(true);
+  const [nodesData, nodesLoaded] = useK8sWatchResource<NodeKind[]>(nodeResource);
+
+  const nodesDataMemoized: NodeKind[] = useDeepCompareMemoize(nodesWithoutTaints(nodesData), true);
+  const uniqZones: Set<string> = new Set(nodesDataMemoized.map((node) => getZone(node)));
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx:89]:
```suggestion
  const uniqZones: Set<string> = new Set(nodesDataMemoized.map(getZone));
```

**Review by bipuladh — CHANGES_REQUESTED** [review]

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/review-and-create.tsx [hunk 8]
```diff
@@ -57,11 +60,22 @@ export const ReviewAndCreate: React.FC<ReviewAndCreateProps> = ({
           <Trans t={t} ns="ceph-storage-plugin" values={osdSize}>
             Requested Cluster Capacity:&nbsp;
             <span className="text-secondary">
-              {{ osdSize }} TiB&nbsp;
+              {osdSize.toString()} TiB&nbsp;
```
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/review-and-create.tsx:63]: Is this necessary ?

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx [hunk 9]
```diff
@@ -42,6 +59,110 @@ const validate = (scName, enableMinimal, enableFlexibleScaling): ValidationType[
...
+  React.useEffect(() => {
+    if (nodesLoaded && enableStretchCluster) {
+      setZonesOptions(_.zipObject([...uniqZones], [...uniqZones]));
+      dispatch({ type: ActionType.SET_ARBITER_ZONE, payload: [...uniqZones]?.[0] });
+    }
+  }, [dispatch, enableStretchCluster, nodesLoaded, uniqZones]);
+  return (
+    <FormGroup fieldId="arbiter-cluster" label={t('ceph-storage-plugin~Stretch Cluster')}>
+      <Checkbox
+        aria-label={t('ceph-storage-plugin~Enable Arbiter')}
```
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx:110]: You already have `label` set , its redundant to add `aria-label` here and with the same text.
```suggestion
```

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/reducer.ts [hunk 10]
```diff
@@ -56,6 +65,9 @@ export const initialState: InternalClusterState = {
   nodes: [],
   enableMinimal: false,
   enableFlexibleScaling: false,
+  enableStretchCluster: false,
+  selectedArbiterZone: '',
+  isArbiterValid: false,
```
Comment (afreen23) [inline frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/reducer.ts:70]: helps to identify the usage of states. Lets follow the same for arbiter.
```suggestion
// arbiter state
  enableStretchCluster: false,
  selectedArbiterZone: '',
  isArbiterValid: false,
```

**Review by afreen23 — CHANGES_REQUESTED** [review]

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/review-and-create.tsx (hunk 8 above)
Comment (cloudbehl) [reply frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/review-and-create.tsx:63]: Yes, it's wasn't working. It was saying `TiB * 3 = 6TiB`  the number was not present

File: frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx (hunk 6 above)
Comment (cloudbehl) [reply frontend/packages/ceph-storage-plugin/src/components/ocs-install/internal-mode/install-wizard-steps/select-capacity-and-nodes-step.tsx]: eslint throws and error when you do the assignment to function parameter.

`Assignment to function parameter 'acc' no-param-reassign`

**Pull request closed without merging**